File: source/packages/@ada/query-parse-render-lambdas/src/test/java/com/ada/query/parserender/utils/api/ApiResponseTest.java
/*! Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */

package com.ada.query.parserender.utils.api;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Modifier;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ApiResponseTest {

    @Test
    public void buildResponseEvent_catches_exception_on_unserializable_input() throws Exception {

        class ClassThatJacksonCannotSerialize {
            final ClassThatJacksonCannotSerialize self = this;

            @Override
            public String toString() {
                return self.getClass().getName();
            }
        }

        ClassThatJacksonCannotSerialize unserializeable = new ClassThatJacksonCannotSerialize();
        APIGatewayProxyResponseEvent response = ApiResponse.success(unserializeable);
        Assertions.assertEquals(500, response.getStatusCode());
        Assertions.assertEquals("{\"message\": \"Failed to serialise response\"}", response.getBody());
    }

    @Test
    public void apiresponse_utility_class_cannot_be_instantiated() throws Exception {
        Constructor<ApiResponse> constructor = ApiResponse.class.getDeclaredConstructor();
        Assertions.assertTrue(Modifier.isPrivate(constructor.getModifiers()));
        constructor.setAccessible(true);
        Assertions.assertThrows(InvocationTargetException.class, () -> {
            constructor.newInstance((Object[]) null);
        });
    }
}

File: source/packages/@ada/infra/src/services/api/api-gateway/__tests__/api-response.test.ts
/*! Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { StatusCodes } from 'http-status-codes';
import { ApiResponse } from '../api-response';

describe('api-response', () => {
  describe('success', () => {
    it('should return 200 and stringify the body', () => {
      const result = ApiResponse.success({ value: 'test' }).asApiGatewayResult();
      expect(result.statusCode).toBe(200);
      expect(JSON.parse(result.body)).toEqual({ value: 'test' });
    });
  });

  describe('badRequest', () => {
    it('should return 400 and stringify the body', () => {
      const result = ApiResponse.badRequest({ message: 'test' }).asApiGatewayResult();
      expect(result.statusCode).toBe(400);
      expect(JSON.parse(result.body)).toEqual({
        name: 'Error',
        message: 'test',
        errorId: expect.stringMatching(/\w{10}/),
      });
    });
  });
  describe('internalServerError', () => {
    it('should return 409 (conflict) and stringify the body with errorId', () => {
      const result = ApiResponse.internalServerError({ message: 'test' }).asApiGatewayResult();
      expect(result.statusCode).toBe(StatusCodes.CONFLICT);
      expect(JSON.parse(result.body)).toEqual({
        name: 'Error',
        message: 'Failed to service request',
        details: expect.stringMatching(/\w{10}/),
        errorId: expect.stringMatching(/\w{10}/),
      });
    });
  });
});
